### DISCUSSION TITLE
eprosima::fastdds::dds::DataReaderListener::on_data_available() with const qualifier

### DISCUSSION THREAD

**Discussion by manuelValch** [eProsima/Fast-DDS#4419] [Support]
Hello, 

trying to call `on_data_available()` method directly from my application (instead of calling it from fastdds thread and context).

Though this method does not have a const qualifier and thus can not being called on a Listener obtained via `dataReader.get_listener()` (which return const).
I can still bypass the issue (with some dirty tricks ctype casts).

Though I wonder if this should be translated as an issue or if there is a cleaner way to call DataReaderListener callbacks ?

**Comment by EduPonz** [comment]
Hi @manuelValch,

I'm wondering what's your use case; `on_data_available()` is meant to be called from the library, so I'd like to understand why'd you like to call it in your application code.

**Reply by manuelValch** [reply]
Hello @EduPonz trying to acheive this to call it from an applicative thread. 
Mostly to maintien compatibility with some legagcy librairies I have. 
Question here is:
Why do `dataReader.get_listener()` needs to return a const ?
Can `on_data_available()` work on a const listener ?

**Reply by EduPonz** [reply]
`get_listener()` returns a `const` because the retriever is not supposed to modify that listener instance. The contract is that from the moment the entity is created, the ownership and management of the listener belongs to the entity, and that the application is always in charge of the lifecycle of the listener object (it needs to create it and delete it), but is is not meant to use it.

That being said, it seems to me that what you want to achieve is always possible, as the application still holds a reference to the listener, so you wouldn't need to access to it via `get_listener()`. I guess if you really know what you're doing, you could use that reference to call to the listener methods, although I'm still unclear as to why you want to do that.

**Reply by manuelValch** [reply]
Hello @EduPonz thank for the answer now it is more clear to me.

Though why does the `DataReaderListener` callbacks are not const qualified ?

That would allow me to call them on the listener obtained from `get_listener()`.

**Reply by EduPonz** [reply]
Well, if the `DataReaderListener` methods were const qualified, it'd mean that the implementation could not do any operation with the `DataReader` in them. That'd mean that an application could not call `take()` not `take_next_sample()` within the `on_data_available()` as those modify the reader's state.

I still fail to understand why you want to call those methods in the application code, but it seems to me that there should be another way around it.

**Reply by manuelValch** [reply]
The main reason of why am I calling those callback from application code is because I need those callbacks to be called in the applicative context not fastdds context. 

Sorry If I am not being clear I would like to keep the possibility of modification on the dataReader though 
i wonder why those callback can not be called on a const subscriber ie
`void on_data_available(fdds::DataReader* pReader) const`

Because subscriber is not supposed to be modifed by the application in those callbacks is it ?

**Reply by EduPonz** [reply]
Exactly, if the method is qualified as `const` then you cannot call any non-const qualified method from it

**Reply by manuelValch** [reply]
ok thanks.
